File: web/src/use-cases/prompt-builder.ts
/**
 * L2 Use Case Utility — Pure functions for building LLM prompts.
 * Mirrors Python l2_use_cases.utils.prompt_builder.
 */

import type { SessionTemplate } from '../entities/template';

/** Build the user prompt for a digest cycle. */
export function buildDigestPrompt(
  template: SessionTemplate,
  buffer: string[],
  options: {
    isFinal?: boolean;
    fullTranscript?: string;
    userContext?: string;
  } = {},
): string {
  const { isFinal = false, fullTranscript = '', userContext = '' } = options;
  const newLines = buffer.join('\n');
  const contextSection = userContext.trim()
    ? `User corrections and additions:\n${userContext.trim()}`
    : '';

  if (isFinal) {
    return template.finalUserTemplate
      .replace('{line_count}', String(buffer.length))
      .replace('{new_lines}', newLines)
      .replace('{user_context}', contextSection)
      .replace('{full_transcript}', fullTranscript || '(no full transcript)');
  }

  return template.digestUserTemplate
    .replace('{line_count}', String(buffer.length))
    .replace('{new_lines}', newLines)
    .replace('{user_context}', contextSection);
}

/** Build the user prompt for a quick action. */
export function buildQuickActionPrompt(
  promptTemplate: string,
  digestMarkdown: string,
  recentTranscript: string,
  options: { userContext?: string } = {},
): string {
  const { userContext = '' } = options;
  let result = promptTemplate
    .replace('{digest_markdown}', digestMarkdown || '(no digest yet)')
    .replace('{recent_transcript}', recentTranscript || '(no transcript yet)');

  if (userContext.trim()) {
    result += `\n\nUser corrections and context:\n${userContext.trim()}`;
  }
  return result;
}

/** Build the synthetic user message for conversation compaction. */
export function buildCompactUserMessage(latestMarkdown: string): string {
  return (
    '(Prior conversation compacted) Current session state:\n\n' +
    `${latestMarkdown}\n\n` +
    'Continue analyzing subsequent transcript segments based on this state.'
  );
}

File: src/lazy_take_notes/l2_use_cases/utils/prompt_builder.py
"""Pure functions for building LLM prompts from templates."""

from __future__ import annotations

from lazy_take_notes.l1_entities.template import SessionTemplate


def build_digest_prompt(
    template: SessionTemplate,
    buffer: list[str],
    *,
    is_final: bool = False,
    full_transcript: str = '',
    user_context: str = '',
) -> str:
    """Build the user prompt for a digest cycle."""
    new_lines = '\n'.join(buffer)
    context_section = f'User corrections and additions:\n{user_context.strip()}' if user_context.strip() else ''

    if is_final:
        return template.final_user_template.format(
            line_count=len(buffer),
            new_lines=new_lines,
            user_context=context_section,
            full_transcript=full_transcript or '(no full transcript)',
        )
    return template.digest_user_template.format(
        line_count=len(buffer),
        new_lines=new_lines,
        user_context=context_section,
    )


def build_quick_action_prompt(
    prompt_template: str,
    digest_markdown: str,
    recent_transcript: str,
    *,
    user_context: str = '',
) -> str:
    """Build the user prompt for a quick action."""
    result = prompt_template.format(
        digest_markdown=digest_markdown or '(no digest yet)',
        recent_transcript=recent_transcript or '(no transcript yet)',
    )
    if user_context.strip():
        result += f'\n\nUser corrections and context:\n{user_context.strip()}'
    return result


def build_label_prompt(template_name: str, template_description: str, latest_digest: str) -> str:
    """Build a prompt that asks the LLM for a short session label."""
    header = f'Template: {template_name}'
    if template_description:
        header += f' — {template_description}'
    return (
        'Generate a short label (2-5 words, snake_case, lowercase) that '
        'summarises this session.\n\n'
        f'{header}\n\n'
        f'Latest digest:\n{latest_digest}\n\n'
        'Output ONLY the label, nothing else.'
    )


def build_compact_user_message(latest_markdown: str) -> str:
    """Build the synthetic user message for conversation compaction."""
    return (
        '(Prior conversation compacted) Current session state:\n\n'
        f'{latest_markdown}\n\n'
        'Continue analyzing subsequent transcript segments based on this state.'
    )
